### PR TITLE
bundle: Suport OLM installation modes properly

### PR DESCRIPTION
**Pull request by jhrozek**
#### What type of PR is this?
/kind bug

#### What this PR does / why we need it:
Please see the commit mmessages for full information about each change.

But the short version is that our OLM packaging was done incorrectly
(and I wrote it..sorry...). The installation mode was ignored unless
the `RESTRICT_TO_NAMESPACE` variable was set and even then installing
SPO using OLM to another namespace than the default one didn't work.

The patchset uses the `olm.targetNamespace` variable set by OLM as
value of the `WATCH_NAMESPACES` variable, but to keep backwards compatibility,
`RESTRICT_TO_NAMESPACE` still takes precedence.

Next, all installation methods that OLM supports are enabled in the bundle.

OLM also manages Service Accounts when SPO is installed through OLM to allow
for installing into non-default namespaces.

Our examples suggest to use the `AllNamespaces` mode as the default, because
by default we want to watch all namespaces but even more importantly,
unless you want to watch a single namespace only, SPO is not multi-tenant
and installing into all namespaces would ensure that only one CSV exists in
the cluster.

Finally, the OLM tests have been expanded quite a bit to cover all different
installation methods.

#### Which issue(s) this PR fixes:
None

#### Does this PR have test?
Yes

#### Special notes for your reviewer:
N/A

#### Does this PR introduce a user-facing change?

```release-note
When using OLM to install the SPO from a bundle, SPO now defaults to installation in all namespaces and watching CRs across all namespaces. Please refer to https://olm.operatorframework.io/docs/advanced-tasks/operator-scoping-with-operatorgroups/ to learn how to scope the operator to either watch only a subset of namespaces or install SPO to a different namespace when using OLM.

Note that the other installation methods or the RESTRICT_TO_NAMESPACE environment variables are not affected by this change and work as before.
```

### CODE REVIEW

File: examples/olm/install-resources.yaml [hunk 1]
```diff
@@ -36,3 +30,9 @@ spec:
   source: security-profiles-operator
   #sourceNamespace: openshift-marketplace on OCP
   sourceNamespace: olm
+---
+apiVersion: operators.coreos.com/v1
+kind: OperatorGroup
+metadata:
+  name: security-profiles-operator
+  namespace: security-profiles-operator
```
Comment (jhrozek) [inline examples/olm/install-resources.yaml:38]: In case anyone is wondering, I just moved the OG to the end of the file so that it's easier to append to in the tests, as a byproduct we can keep using this example in the test which would hopefully also keep testing the example.

File: hack/ci/e2e-olm.sh [hunk 2]
```diff
@@ -50,30 +63,112 @@ function build_and_push_packages() {
...
+    ns=$1
+
     # Useful in case the CatalogSource is fubar. We retry several times
     # because on transient errors (which are for some reason common even
     # if the catalog is local) the pod gets restarted
     for i in $(seq 1 5); do
         kubectl -nolm wait --for=condition=ready pods -lolm.catalogSource=security-profiles-operator
-        catalog_logs=$(kubectl -nolm logs $(kubectl -nolm get pods --no-headers -lolm.catalogSource=security-profiles-operator | awk '{print $1}') 2>/dev/null)
-        if [[ -n "$catalog_logs" ]]; then
+        install_rv=$?
```
Comment (jhrozek) [inline hack/ci/e2e-olm.sh:166]: This change is not strictly related to the tests, but I found (because I was running the tests over and over..) that the wait occasionally times out on the first try, but the `catalog_logs` would usually contain something, so I made the retry loop on `install_rv` instead.

File: hack/ci/e2e-olm.sh [hunk 3]
```diff
@@ -82,30 +177,212 @@ function check_spo_is_running() {
     # (jhrozek): I didn't find a useful condition or status to wait for..
     # ..if only there was a way to check if ANY installedCSV is set..
     sleep 30
-    CSV=$(kubectl -nsecurity-profiles-operator get sub security-profiles-operator-sub -ojsonpath='{.status.installedCSV}')
+    CSV=$(kubectl -n$ns get sub security-profiles-operator-sub -ojsonpath='{.status.installedCSV}')
     # wait for the CSV to be actually installed
-    kubectl -nsecurity-profiles-operator wait --for=jsonpath='{.status.phase}'=Succeeded csv $CSV
+    kubectl -n$ns wait --for=jsonpath='{.status.phase}'=Succeeded csv $CSV
```
Comment (jhrozek) [inline hack/ci/e2e-olm.sh:182]: `$ns` instead if `security-profiles-operator` to support the custom namespace test.

File: hack/ci/e2e-olm.sh [hunk 4]
```diff
@@ -82,30 +177,212 @@ function check_spo_is_running() {
...
-    kubectl -nsecurity-profiles-operator wait --for=condition=ready pod -lname=spod
+    kubectl -n$ns wait --for=condition=ready pod -lname=spod || return 1
+
+    return 0
+}
+
+function assert_spo_csv_installed_in_ns() {
+    ns=$1
+
+    [[ $(kubectl get csv -loperators.coreos.com/security-profiles-operator.$ns= -n$ns -oname) ]] || return 1
```
Comment (jhrozek) [inline hack/ci/e2e-olm.sh:217]: this probably looks hacky, but I found it easier to list any CSV in the NS that installs SPO rather than parse out the version of the CSV directly..
the `-loperators.coreos.com/security-profiles-operator.$ns=` is an OLM-ism, the label is `-loperators.coreos.com/$NAME_OF_THE_OPERATOR.$NAMESPACE_WHERE_THE_OPERATOR_IS_INSTALLED=`

File: hack/ci/e2e-olm.sh [hunk 5]
```diff
@@ -82,30 +177,212 @@ function check_spo_is_running() {
...
+    smoke_test $method || rv=1
+
+    teardown_spo $method
+done
+
+# deployment into the custom namespace is a bit special
+echo "Testing SPO deployment in custom namespace(s)"
+deploy_spo_in_custom_ns spo-lives-here || rv=1
+check_spo_is_running spo-lives-here || rv=1
+kubectl get csv -A --show-labels
```
Comment (jhrozek) [inline hack/ci/e2e-olm.sh:384]: hmm, sorry, this is not needed and is just debugging, I can either remove this command if there's comment for this PR or when I submit a patch for the environment variable installations.

**Review by saschagrunert — APPROVED** [review]

File: hack/ci/e2e-olm.sh (hunk 5 above)
Comment (saschagrunert) [reply hack/ci/e2e-olm.sh:384]: Let's keep it if it helps debugging.
